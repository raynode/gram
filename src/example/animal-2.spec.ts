import {
  GraphQLBoolean,
  GraphQLInt,
  GraphQLString,
  printSchema,
} from 'graphql'
import { createSchemaBuilder, Paged, Service } from '..'

enum AnimalTypes {
  Cat,
  Dog,
}

interface Animal {
  type: AnimalTypes
  name: string
  tame: boolean
  age: number
}

interface Cat extends Animal {
  type: AnimalTypes.Cat
}
interface Dog extends Animal {
  type: AnimalTypes.Dog
}

const Animals: Service<Animal> = {
  findOne: async () => null,
  findMany: async () => ({
    page: null,
    nodes: [],
  }),
}

const createAnimalService = <Type extends Animal>(
  type: AnimalTypes,
): Service<Type> => ({
  findOne: async ({ order, where }, context) =>
    Animals.findOne(
      {
        order,
        where: { ...where, type },
      },
      context,
    ) as Promise<Type>,
  findMany: async ({ order, page, where }, context) =>
    Animals.findMany(
      {
        order,
        page,
        where: { ...where, type },
      },
      context,
    ) as Promise<Paged<Type>>,
})

const Cats = createAnimalService<Cat>(AnimalTypes.Cat)
const Dogs = createAnimalService<Cat>(AnimalTypes.Dog)

describe('testing the example 2', () => {
  it('should build the example code', () => {
    const builder = createSchemaBuilder()
    const animal = builder.interface('Animal')
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('age', GraphQLInt)

    const cat = builder.model('Cat', Animals)
    cat.interface('Animal')

    expect(printSchema(builder.build())).toMatchSnapshot()
  })

  it('should build the second part of the example', () => {
    const builder = createSchemaBuilder()
    const animal = builder.interface('Animal', Animals)
    animal.attr('name', GraphQLString)
    animal.attr('tame', GraphQLBoolean)
    animal.attr('age', GraphQLInt)

    const cat = builder.model('Cat', Cats)
    cat.interface('Animal')
    const dog = builder.model('Dog', Dogs)
    dog.interface('Animal')

    expect(printSchema(builder.build())).toMatchSnapshot()
  })
})
