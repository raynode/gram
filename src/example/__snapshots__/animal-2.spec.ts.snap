// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 2 should build the example code 1`] = `
"interface Animal {
  name: String
  tame: Boolean
  age: Int
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

type Cat implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input CatFilter {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input CatPage {
  limit: Int
  offset: Int
}

type Cats implements List {
  page: Page
  nodes: [Cat!]!
}

enum CatSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input CatWhere {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input CreateAnimalData {
  name: String
  tame: Boolean
  age: Int
}

input CreateCatData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getCat(where: CatWhere!, order: CatSortOrder): Cat
  getCats(order: CatSortOrder, page: CatPage, where: CatWhere!): Cats
}

input UpdateAnimalData {
  name: String
  tame: Boolean
  age: Int
}

input UpdateCatData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}
"
`;

exports[`testing the example 2 should build the second part of the example 1`] = `
"interface Animal {
  name: String
  tame: Boolean
  age: Int
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

type Cat implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input CatFilter {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input CatPage {
  limit: Int
  offset: Int
}

type Cats implements List {
  page: Page
  nodes: [Cat!]!
}

enum CatSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input CatWhere {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input CreateAnimalData {
  name: String
  tame: Boolean
  age: Int
}

input CreateCatData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input CreateDogData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

scalar DateTime

type Dog implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input DogFilter {
  AND: [DogFilter!]
  OR: [DogFilter!]
  NOT: DogFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

input DogPage {
  limit: Int
  offset: Int
}

type Dogs implements List {
  page: Page
  nodes: [Dog!]!
}

enum DogSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input DogWhere {
  AND: [DogFilter!]
  OR: [DogFilter!]
  NOT: DogFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

interface List {
  page: Page
  nodes: [Node!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getAnimals(order: AnimalSortOrder, page: AnimalPage, where: AnimalWhere!): Animals
  getCat(where: CatWhere!, order: CatSortOrder): Cat
  getCats(order: CatSortOrder, page: CatPage, where: CatWhere!): Cats
  getDog(where: DogWhere!, order: DogSortOrder): Dog
  getDogs(order: DogSortOrder, page: DogPage, where: DogWhere!): Dogs
}

input UpdateAnimalData {
  name: String
  tame: Boolean
  age: Int
}

input UpdateCatData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input UpdateDogData {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}
"
`;
